keps/sig-instrumentation/647-apiserver-tracing/README.md

```diff
@@ -108,52 +132,37 @@ The [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-co
-  Service *ServiceReference `json:"service,omitempty" protobuf:"bytes,1,opt,name=service"`
-}
+  // URL of the collector that's running on the control-plane node.
+  // the APIServer uses the egressType ControlPlane when sending data to the collector.
+  // Defaults to localhost:4317
+  URL *string `json:"url,omitempty" protobuf:"bytes,1,opt,name=url"`
 
-// ServiceReference holds a reference to Service.legacy.k8s.io
-type ServiceReference struct {
-  // `namespace` is the namespace of the service.
-  // Required
-  Namespace string `json:"namespace" protobuf:"bytes,1,opt,name=namespace"`
-  // `name` is the name of the service.
-  // Required
-  Name string `json:"name" protobuf:"bytes,2,opt,name=name"`
-
-  // If specified, the port on the service.
-  // Defaults to 4317, the IANA reserved port for OpenTelemetry.
-  // `port` should be a valid port number (1-65535, inclusive).
   // +optional
-  Port *int32 `json:"port,omitempty" protobuf:"varint,3,opt,name=port"`
+  // SamplingRatePerMillion is the number of samples to collect per million spans.
+  // Defaults to 0.
+  SamplingRatePerMillion *int32 `json:"samplingRatePerMillion,omitempty" protobuf:"varint,2,opt,name=samplingRatePerMillion"`
 }
 ```
 
-If `--opentelemetry-config-file` is not specified, the API Server will not send any telemetry.
+If `--opentelemetry-config-file` is not specified, the API Server will not send any spans, even if incoming requests are sampled.
 
 ### Controlling use of the OpenTelemetry library
 
 As the community found in the [Metrics Stability Framework KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/1209-metrics-stability/kubernetes-control-plane-metrics-stability.md#kubernetes-control-plane-metrics-stability), having control over how the client libraries are used in kubernetes can enable maintainers to enforce policy and make broad improvements to the quality of telemetry.  To enable future improvements to tracing, we will restrict the direct use of the OpenTelemetry library within the kubernetes code base, and provide wrapped versions of functions we wish to expose in a utility library.
 
 ### Test Plan
 
-We will e2e test this feature by deploying an OpenTelemetry Collector on the master, and configure it to export traces using the [stdout exporter](https://github.com/open-telemetry/opentelemetry-go/tree/master/exporters/stdout), which logs the spans in json format.  We can then verify that the logs contain our expected traces when making calls to the API Server.
+We will e2e test this feature by deploying an OpenTelemetry Collector on the control-plane node, and configure it to export traces using the [stdout exporter](https://github.com/open-telemetry/opentelemetry-go/tree/master/exporters/stdout), which logs the spans in json format.  We can then verify that the logs contain our expected traces when making calls to the API Server.
```

Comment:
In a manual or automated test? This seems very heavyweight for an automated test, but maybe I'm wrong.

Reply:
The current plan is for it to be an automated test.  The protocol is pushed-based, so we can't just query the traces like we can for prometheus metrics.  If we want to check that traces are being generated, we need to run something where the apiserver is pushing the traces.

Do you have any suggestions for an alternative approach?  I'd prefer to have _some_ automated testing if at all possible.

Reply:
OK, I don't know if I have a better idea. Reading the logs just sounds like an enormous pain.

Reply:
There are in-memory providers suited for testing, but I agree that "e2e" sounds like wiring up an exporter of some kind.

https://pkg.go.dev/go.opentelemetry.io/otel/oteltest#hdr-Trace_Testing